### ISSUE TITLE
<?xml version="1.0"?> Error when trying to create customer/invoice

### ISSUE THREAD

**Issue by djskim27** [ruckus/quickbooks-ruby#421]
Is anyone else experiencing this error unexpectedly all of a sudden? I'm using the o-auth-2 branch and everything's been working smoothly until today I started getting this error when trying to create a new customer.

Any tips on how I could troubleshoot this?

**Comment by ruckus** [comment]
Can you please enable logging and provide the request/response?

https://github.com/ruckus/quickbooks-ruby#logging

**Comment by djskim27** [comment]
I apologize for not providing the log earlier!

https://i.imgur.com/dKfjYD1.png

Please let me know if you prefer me copying and pasting the actual log contents on here instead.

**Comment by djskim27** [comment]
Request: https://i.imgur.com/yhMmoAk.png
Response: https://i.imgur.com/gtiB5VV.png

It seems every post request that I make to the quickbooks api is consistently returning this error. Any advice? I'm puzzled because everything was working seamlessly before today.

**Comment by ruckus** [comment]
I see whats happening. I'm working on a fix.

**Comment by ruckus** [comment]
Ok just pushed a commit. Wanna try `389-oauth2` HEAD?

**Comment by djskim27** [comment]
I am still receiving the same xml error. I pulled from your latest commit.

**Comment by ruckus** [comment]
Hmm, its working for.

```
[cody:webapp (389-oauth2)]$ ruby client.rb
I, [2018-03-14T20:10:10.414741 #88445]  INFO -- : ------ QUICKBOOKS-RUBY REQUEST ------
I, [2018-03-14T20:10:10.414817 #88445]  INFO -- : METHOD = post
I, [2018-03-14T20:10:10.414834 #88445]  INFO -- : RESOURCE = https://sandbox-quickbooks.api.intuit.com/v3/company/1313827825/customer
I, [2018-03-14T20:10:10.414848 #88445]  INFO -- : REQUEST BODY:
I, [2018-03-14T20:10:10.414982 #88445]  INFO -- : <?xml version="1.0" encoding="utf-8"?>
<Customer xmlns="http://schema.intuit.com/finance/v3" sparse="false">
<CompanyName>Test 123-a714f2</CompanyName>
<DisplayName>Test 123-a714f2</DisplayName>
<PrimaryEmailAddr>
<Address>test@example.com</Address>
</PrimaryEmailAddr>
</Customer>

I, [2018-03-14T20:10:10.415022 #88445]  INFO -- : REQUEST HEADERS = {"Content-Type"=>"application/xml", "Accept"=>"application/xml", "Accept-Encoding"=>"gzip, deflate"}
I, [2018-03-14T20:10:10.828863 #88445]  INFO -- : ------ QUICKBOOKS-RUBY RESPONSE ------
I, [2018-03-14T20:10:10.828942 #88445]  INFO -- : RESPONSE CODE = 200
I, [2018-03-14T20:10:10.829032 #88445]  INFO -- : RESPONSE HEADERS = {"server"=>"nginx", "date"=>"Thu, 15 Mar 2018 03:10:10 GMT", "content-type"=>"application/xml;charset=UTF-8", "content-length"=>956, "strict-transport-security"=>"max-age=15552000", "intuit_tid"=>"e9966f5a-edf2-a3e5-9384-f6253cb39298", "x-content-type-options"=>"nosniff", "set-cookie"=>"JSESSIONID=688F38D3B4D4EE6520E1742356C61DD0.c50-pprdsbxas901; Domain=qbo.intuit.com; Path=/; Secure; HttpOnly", "qbo-version"=>"1802.1292", "cache-control"=>"max-age=0, no-cache, no-store, must-revalidate, private", "service-time"=>"total=131, db=43", "expires"=>"0", "connection"=>"Close"}
I, [2018-03-14T20:10:10.829069 #88445]  INFO -- : RESPONSE BODY:
I, [2018-03-14T20:10:10.829322 #88445]  INFO -- : <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<IntuitResponse xmlns="http://schema.intuit.com/finance/v3" time="2018-03-14T20:10:10.659-07:00">
  <Customer domain="QBO" sparse="false">
    <Id>76</Id>
    <SyncToken>0</SyncToken>
    <MetaData>
      <CreateTime>2018-03-14T20:10:10-07:00</CreateTime>
      <LastUpdatedTime>2018-03-14T20:10:10-07:00</LastUpdatedTime>
    </MetaData>
    <FullyQualifiedName>Test 123-a714f2</FullyQualifiedName>
    <CompanyName>Test 123-a714f2</CompanyName>
    <DisplayName>Test 123-a714f2</DisplayName>
    <PrintOnCheckName>Test 123-a714f2</PrintOnCheckName>
    <Active>true</Active>
    <PrimaryEmailAddr>
      <Address>test@example.com</Address>
    </PrimaryEmailAddr>
    <DefaultTaxCodeRef>2</DefaultTaxCodeRef>
    <Taxable>true</Taxable>
    <Job>false</Job>
    <BillWithParent>false</BillWithParent>
    <Balance>0</Balance>
    <BalanceWithJobs>0</BalanceWithJobs>
    <CurrencyRef name="United States Dollar">USD</CurrencyRef>
    <PreferredDeliveryMethod>Print</PreferredDeliveryMethod>
  </Customer>
</IntuitResponse>
```

**Comment by djskim27** [comment]
https://i.imgur.com/whKNjtQ.png

It says I have a xml syntax error (undefined namespace prefix).

Do I need to pull an updated version of nokogiri? or Is there something wrong in the syntax of my request?

**Comment by ruckus** [comment]
Your error is what I was getting before I implemented the fix. Are you sure you are using this commit with the gzip decompression? 

https://github.com/ruckus/quickbooks-ruby/commit/49cbb81aa347821323c58f8f3e103c37e11485d3

**Comment by ruckus** [comment]
Hmm, yeah that Gemfile command looks correct. Try wiping your `Gemfile.lock` and trying again. Perhaps bundler is using a cached version and is not actually updating and pulling down the latest commit.

After you do that lets verify that it did indeed pull down the latest by looking for the presence of the `lib/faraday/middleware/gzip.rb` file like this:

```bash
$ gem which quickbooks-ruby
/Users/cody/.rvm/gems/ruby-2.1.1@cru-rails3-aws/gems/quickbooks-ruby-0.5.1/lib/quickbooks-ruby.rb
$ cd /Users/cody/.rvm/gems/ruby-2.1.1@cru-rails3-aws/gems/quickbooks-ruby-0.5.1
$ ls -R lib/faraday/middleware
```

And you should see some files. If that directory (`lib/faraday`) does *not* exist then your bundler is not pulling down the latest

**Comment by djskim27** [comment]
Thank you so much for your patience and willingness to help me out..

I tried wiping the Gemfile.lock and I reran 'bundle install'

It still does not appear to be pulling the correct commit for the 389-oauth2 branch still:

"Using quickbooks-ruby 0.6.1 from https://github.com/ruckus/quickbooks-ruby.git (at 389-oauth2@c21910b)"

lib/faraday also does not appear.. so my bundler is not pulling down the latest commit :(

**Comment by ruckus** [comment]
Ok, so I just learned that when you use `:git` in Gemfile it does not place the gem in where `gem show` would find it. But we can ask Bundler to show us where it stores it:

When I do this in a Rails project:

```
gem 'quickbooks-ruby', git: 'https://github.com/ruckus/quickbooks-ruby.git', branch: "389-oauth2"
```

I can use bundler to see the checked out code:

```
$ bundle show quickbooks-ruby
/Users/cody/.rvm/gems/ruby-2.3.1@strawberryshare/bundler/gems/quickbooks-ruby-c21910b44901

$ cd `bundle show quickbooks-ruby`
$ ls lib/quickbooks/faraday/
```

What I think is happening is you have multiple copies of them gem and its getting confused. I think just wiping `Gemfile.lock` is not sufficient. You might need to wipe everything under the `/Users/cody/.rvm/gems/ruby-2.3.1@strawberryshare/bundler/gems` directory (obvi yours will be different). Also uninstall any quickbooks-ruby gems via

```
$ gem uninstall quickbooks-ruby
```

Basically try starting with a clean slate

**Comment by djskim27** [comment]
so I followed your instructions and cleared everything (Gemfile.lock and Bundler Gems included).
I reran bundle install and searched for the faraday/middleware/gzip.rb and found it. I turned on my server and tried to create a customer within quickbooks again. I received the same error :(

So either I am still not pulling the updated files or something else is wrong with my setup. Would it be too much to ask if I could try pulling these changes from a cloned branch of a different name? Or would that even change anything?

**Comment by djskim27** [comment]
ah nvm, I clearly did NOT clear every gem file in my bundler gems directory. You were 100 percent correct about the multiple versions of quickbooks ruby being saved. I will do clear this correctly and test it one more time.

Thank you so much for your in-depth instructions and help!

**Comment by djskim27** [comment]
Okay, so I cleared everything correctly and pulled in the files directly from the commit 49cbb81 by referencing 'ref: "49cbb81" and running bundle install.

I found the farraday gzip file to verify that I was pulling the correct changes.. and retried Quickbooks and I am still receiving the original error. At this point, I am ready to throw in the towel unless you have any other suggestions?

I can't thank you enough though for consistently responding to me!

**Comment by brianhenryhf** [comment]
just to make absolutely sure, if you want to, these tricks might be helpful to make sure your app is using what you expect:

- if you're working with a rails app, do `bundle exec rails c` and then exec `require 'rubygems'; Gem::Specification.select {|it| it.name =~ /quickbooks.+/}.each {|it| p it; puts '--';};nil` to dump what rubygems is using for the gem info.  I think `loaded_from` should show where it's pulling the gem (and this will have a SHA for the commit you're using).  who knows, might have some other version or something else bringing in the gem as a dependency with a locked version or whatever.  call it a sanity check.

- similarly, in console, run `puts Quickbooks::Service::AccessToken.instance_method(:renew).source_location` - this should show where on the filesystem your app is sourcing that quickbooks gem resource (can work back to root of gem from there).

if you're not working with a rails app, just cram that stuff somewhere in your app, hit that code and watch logs.

**Comment by djskim27** [comment]
thank you brian! I did what you did and was able to pull the latest commit. However I am still running into a syntax error with XML...

```

I, [2018-03-15T17:42:00.602347 #6614]  INFO -- : ------ QUICKBOOKS-RUBY REQUEST ------
I, [2018-03-15T17:42:00.602436 #6614]  INFO -- : METHOD = post
I, [2018-03-15T17:42:00.602505 #6614]  INFO -- : RESOURCE = https://sandbox-quickbooks.api.intuit.com/v3/company/193514718332604/customer
I, [2018-03-15T17:42:00.602570 #6614]  INFO -- : REQUEST BODY:
I, [2018-03-15T17:42:00.602745 #6614]  INFO -- : <?xml version="1.0" encoding="utf-8"?>
<Customer xmlns="http://schema.intuit.com/finance/v3" sparse="false">
<Title/>
<GivenName>Doreen</GivenName>
<MiddleName/>
<FamilyName>Chang</FamilyName>
<DisplayName>Doreen Chang(0)</DisplayName>
<PrimaryPhone>
<FreeFormNumber>(678) 687-0906</FreeFormNumber>
</PrimaryPhone>
<PrimaryEmailAddr>
<Address>chang@chang.com</Address>
</PrimaryEmailAddr>
<BillAddr>
<Line1>123 Chang Street</Line1>
<City>Atlanta</City>
<CountrySubDivisionCode>GA</CountrySubDivisionCode>
<PostalCode>30024</PostalCode>
</BillAddr>
<Suffix/>
</Customer>

I, [2018-03-15T17:42:00.603430 #6614]  INFO -- : REQUEST HEADERS = {"Content-Type"=>"application/xml", "Accept"=>"application/xml", "Accept-Encoding"=>"gzip, deflate"}
I, [2018-03-15T17:42:02.142224 #6614]  INFO -- : ------ QUICKBOOKS-RUBY RESPONSE ------
I, [2018-03-15T17:42:02.142407 #6614]  INFO -- : RESPONSE CODE = 200
I, [2018-03-15T17:42:02.142540 #6614]  INFO -- : RESPONSE HEADERS = {"server"=>"nginx", "date"=>"Thu, 15 Mar 2018 17:42:02 GMT", "content-type"=>"application/xml;charset=UTF-8", "content-length"=>"617", "connection"=>"close", "strict-transport-security"=>"max-age=15552000", "intuit_tid"=>"64a53043-c189-a218-168b-e4cffde7b730", "x-content-type-options"=>"nosniff", "set-cookie"=>"JSESSIONID=6C6021A5F7B3288DD1FFEF62D0C08AED.c50-pprdsbxas902; Domain=qbo.intuit.com; Path=/; Secure; HttpOnly", "qbo-version"=>"1802.1355", "cache-control"=>"max-age=0, no-cache, no-store, must-revalidate, private", "expires"=>"0", "service-time"=>"total=985, db=787", "content-encoding"=>"gzip"}
I, [2018-03-15T17:42:02.142641 #6614]  INFO -- : RESPONSE BODY:
I, [2018-03-15T17:42:02.142889 #6614]  INFO -- : <?xml version="1.0"?>

WTF: #<Nokogiri::XML::XPath::SyntaxError: ERROR: Undefined namespace prefix: //xmlns:IntuitResponse/xmlns:Fault>
There was an issue creating the user on quickbooks. Please contact an administrator.
```

I looked into bundler and the gzip.rb file in the faraday directory exists.
```
ubuntu@ip-172-31-28-44:~/.rvm/gems/ruby-2.4.1/bundler/gems/quickbooks-ruby-49cbb81aa347$ ls lib/quickbooks/faraday/middleware
gzip.rb

```

**Comment by djskim27** [comment]
https://i.imgur.com/nzncATG.png

I opened up the gzip.rb and it seems that it matches the changes of your commit @ruckus 

But still no dice with being able to perform a successful post request to quickbooks..

**Comment by ruckus** [comment]
What version of Nokogiri do you have installed?

```
[cody:quickbooks-ruby (389-oauth2)]$ gem list | grep nokogiri
nokogiri (1.8.2, 1.8.1, 1.8.0)
```

In this case `quickbook-ruby` would be using the highest version, so for me its `1.8.2`. But I am curious if you're using a different version

**Comment by djskim27** [comment]
```
ubuntu@ip-172-31-28-44:~/folder/charles-miller-app$ gem list | grep nokogiri
nokogiri (1.8.2, 1.8.1)
```
I'm using 1.8.2 haha.

**Comment by ruckus** [comment]
Hmm, I stumped at this point. I just cannot replicate whats happening to you.

Well, the only way I *can* replicate it is in a way that I know the gem is not doing. But just to describe it.

1. Take your Customer response XML and save it to a file.
2. In a `rails console` or `irb` session load up nokogiri and feed it that XML file
3. Run the XPath query that is related to this exception to be raised:

https://github.com/ruckus/quickbooks-ruby/blob/389-oauth2/lib/quickbooks/service/base_service.rb#L411

4. It runs fine in
5. Now, ask Nogogiri to remove namespaces.
6. Re-run the xpath and it fails with the exception you're seeing

This would make sense *IF* the gem was removing namespaces - but its not. Thus, I just cannot seem to replicate that exception.

Full steps of what I'm talking about.

```
# tmp/customer.xml is your response XML
$ irb
2.2.6 :001 > require 'nokogiri'

2.2.6 :003 > x = Nokogiri::XML(File.read("tmp/customer.xml"))
 => #<Nokogiri::XML::Document:0x3fef558dade0 name="document" children=[#<Nokogiri::XML::Element:0x3fef56027d14 name="Customer" namespace=#<Nokogiri::XML::Namespace:0x3fef56027cb0 href="http://schema.intuit.com/finance/v3"> attributes=[#<Nokogiri::XML::Attr:0x3fef56027c88 name="sparse" value="false">] children=[#<Nokogiri::XML::Text:0x3fef56027670 "\n">, #<Nokogiri::XML::Element:0x3fef560275bc name="Title" namespace=#<Nokogiri::XML::Namespace:0x3fef56027cb0 href="http://schema.intuit.com/finance/v3">>, #<Nokogiri::XML::Text:0x3fef56027224 "\n">, #<Nokogiri::XML::Element:0x3fef56027170 name="GivenName" namespace=#<Nokogiri::XML::Namespace:0x3fef56027cb0 href="http://schema.intuit.com/finance/v3"> children=[#<Nokogiri::XML::Text:0x3fef56026dc4 "Doreen">]>, #<Nokogiri::XML::Text:0x3fef56026c20 "\n">, #<Nokogiri::XML::Element:0x3fef56026b6c name="MiddleName" namespace=#<Nokogiri::XML::Namespace:0x3fef56027cb0 href="http://schema.intuit.com/finance/v3">>, #<Nokogiri::XML::Text:0x3fef560267d4 "\n">, #<Nokogiri::XML::Element:0x3fef56026720 name="FamilyName" namespace=#<Nokogiri::XML::Namespace:0x3fef56027cb0 href="http://schema.intuit.com/finance/v3"> children=[#<Nokogiri::XML::Text:0x3fef56026374 "Chang">]>, #<Nokogiri::XML::Text:0x3fef560261d0 "\n">, #<Nokogiri::XML::Element:0x3fef5602611c name="DisplayName" namespace=#<Nokogiri::XML::Namespace:0x3fef56027cb0 href="http://schema.intuit.com/finance/v3"> children=[#<Nokogiri::XML::Text:0x3fef56421cbc "Doreen Chang(0)">]>, #<Nokogiri::XML::Text:0x3fef56421a78 "\n">, #<Nokogiri::XML::Element:0x3fef56421960 name="PrimaryPhone" namespace=#<Nokogiri::XML::Namespace:0x3fef56027cb0 href="http://schema.intuit.com/finance/v3"> children=[#<Nokogiri::XML::Text:0x3fef56421190 "\n">, #<Nokogiri::XML::Element:0x3fef564210c8 name="FreeFormNumber" namespace=#<Nokogiri::XML::Namespace:0x3fef56027cb0 href="http://schema.intuit.com/finance/v3"> children=[#<Nokogiri::XML::Text:0x3fef56420c2c "(678) 687-0906">]>, #<Nokogiri::XML::Text:0x3fef56420a24 "\n">]>, #<Nokogiri::XML::Text:0x3fef5642081c "\n">, #<Nokogiri::XML::Element:0x3fef56420740 name="PrimaryEmailAddr" namespace=#<Nokogiri::XML::Namespace:0x3fef56027cb0 href="http://schema.intuit.com/finance/v3"> children=[#<Nokogiri::XML::Text:0x3fef56420240 "\n">, #<Nokogiri::XML::Element:0x3fef564200b0 name="Address" namespace=#<Nokogiri::XML::Namespace:0x3fef56027cb0 href="http://schema.intuit.com/finance/v3"> children=[#<Nokogiri::XML::Text:0x3fef5591ddd4 "chang@chang.com">]>, #<Nokogiri::XML::Text:0x3fef5591dc30 "\n">]>, #<Nokogiri::XML::Text:0x3fef5591da8c "\n">, #<Nokogiri::XML::Element:0x3fef5591d9d8 name="BillAddr" namespace=#<Nokogiri::XML::Namespace:0x3fef56027cb0 href="http://schema.intuit.com/finance/v3"> children=[#<Nokogiri::XML::Text:0x3fef5591d62c "\n">, #<Nokogiri::XML::Element:0x3fef5591d578 name="Line1" namespace=#<Nokogiri::XML::Namespace:0x3fef56027cb0 href="http://schema.intuit.com/finance/v3"> children=[#<Nokogiri::XML::Text:0x3fef5591d1cc "123 Chang Street">]>, #<Nokogiri::XML::Text:0x3fef5591d028 "\n">, #<Nokogiri::XML::Element:0x3fef5591cf74 name="City" namespace=#<Nokogiri::XML::Namespace:0x3fef56027cb0 href="http://schema.intuit.com/finance/v3"> children=[#<Nokogiri::XML::Text:0x3fef5591cbc8 "Atlanta">]>, #<Nokogiri::XML::Text:0x3fef5591ca24 "\n">, #<Nokogiri::XML::Element:0x3fef5591c970 name="CountrySubDivisionCode" namespace=#<Nokogiri::XML::Namespace:0x3fef56027cb0 href="http://schema.intuit.com/finance/v3"> children=[#<Nokogiri::XML::Text:0x3fef5591c5c4 "GA">]>, #<Nokogiri::XML::Text:0x3fef5591c420 "\n">, #<Nokogiri::XML::Element:0x3fef5591c36c name="PostalCode" namespace=#<Nokogiri::XML::Namespace:0x3fef56027cb0 href="http://schema.intuit.com/finance/v3"> children=[#<Nokogiri::XML::Text:0x3fef5641df54 "30024">]>, #<Nokogiri::XML::Text:0x3fef5641dc98 "\n">]>, #<Nokogiri::XML::Text:0x3fef5641d9f0 "\n">, #<Nokogiri::XML::Element:0x3fef5641d900 name="Suffix" namespace=#<Nokogiri::XML::Namespace:0x3fef56027cb0 href="http://schema.intuit.com/finance/v3">>, #<Nokogiri::XML::Text:0x3fef5641c94c "\n">]>]>

2.2.6 :004 > x.xpath("//xmlns:IntuitResponse/xmlns:Fault")
 => []

2.2.6 :006 > x.remove_namespaces!
 => #<Nokogiri::XML::Document:0x3fef558dade0 name="document" children=[#<Nokogiri::XML::Element:0x3fef56027d14 name="Customer" attributes=[#<Nokogiri::XML::Attr:0x3fef56027c88 name="sparse" value="false">] children=[#<Nokogiri::XML::Text:0x3fef56027670 "\n">, #<Nokogiri::XML::Element:0x3fef560275bc name="Title">, #<Nokogiri::XML::Text:0x3fef56027224 "\n">, #<Nokogiri::XML::Element:0x3fef56027170 name="GivenName" children=[#<Nokogiri::XML::Text:0x3fef56026dc4 "Doreen">]>, #<Nokogiri::XML::Text:0x3fef56026c20 "\n">, #<Nokogiri::XML::Element:0x3fef56026b6c name="MiddleName">, #<Nokogiri::XML::Text:0x3fef560267d4 "\n">, #<Nokogiri::XML::Element:0x3fef56026720 name="FamilyName" children=[#<Nokogiri::XML::Text:0x3fef56026374 "Chang">]>, #<Nokogiri::XML::Text:0x3fef560261d0 "\n">, #<Nokogiri::XML::Element:0x3fef5602611c name="DisplayName" children=[#<Nokogiri::XML::Text:0x3fef56421cbc "Doreen Chang(0)">]>, #<Nokogiri::XML::Text:0x3fef56421a78 "\n">, #<Nokogiri::XML::Element:0x3fef56421960 name="PrimaryPhone" children=[#<Nokogiri::XML::Text:0x3fef56421190 "\n">, #<Nokogiri::XML::Element:0x3fef564210c8 name="FreeFormNumber" children=[#<Nokogiri::XML::Text:0x3fef56420c2c "(678) 687-0906">]>, #<Nokogiri::XML::Text:0x3fef56420a24 "\n">]>, #<Nokogiri::XML::Text:0x3fef5642081c "\n">, #<Nokogiri::XML::Element:0x3fef56420740 name="PrimaryEmailAddr" children=[#<Nokogiri::XML::Text:0x3fef56420240 "\n">, #<Nokogiri::XML::Element:0x3fef564200b0 name="Address" children=[#<Nokogiri::XML::Text:0x3fef5591ddd4 "chang@chang.com">]>, #<Nokogiri::XML::Text:0x3fef5591dc30 "\n">]>, #<Nokogiri::XML::Text:0x3fef5591da8c "\n">, #<Nokogiri::XML::Element:0x3fef5591d9d8 name="BillAddr" children=[#<Nokogiri::XML::Text:0x3fef5591d62c "\n">, #<Nokogiri::XML::Element:0x3fef5591d578 name="Line1" children=[#<Nokogiri::XML::Text:0x3fef5591d1cc "123 Chang Street">]>, #<Nokogiri::XML::Text:0x3fef5591d028 "\n">, #<Nokogiri::XML::Element:0x3fef5591cf74 name="City" children=[#<Nokogiri::XML::Text:0x3fef5591cbc8 "Atlanta">]>, #<Nokogiri::XML::Text:0x3fef5591ca24 "\n">, #<Nokogiri::XML::Element:0x3fef5591c970 name="CountrySubDivisionCode" children=[#<Nokogiri::XML::Text:0x3fef5591c5c4 "GA">]>, #<Nokogiri::XML::Text:0x3fef5591c420 "\n">, #<Nokogiri::XML::Element:0x3fef5591c36c name="PostalCode" children=[#<Nokogiri::XML::Text:0x3fef5641df54 "30024">]>, #<Nokogiri::XML::Text:0x3fef5641dc98 "\n">]>, #<Nokogiri::XML::Text:0x3fef5641d9f0 "\n">, #<Nokogiri::XML::Element:0x3fef5641d900 name="Suffix">, #<Nokogiri::XML::Text:0x3fef5641c94c "\n">]>]>

2.2.6 :007 > x.xpath("//xmlns:IntuitResponse/xmlns:Fault")
Nokogiri::XML::XPath::SyntaxError: ERROR: Undefined namespace prefix: //xmlns:IntuitResponse/xmlns:Fault
	from /Users/cody/.rvm/gems/ruby-2.2.6@quickbooks-ruby/gems/nokogiri-1.8.2/lib/nokogiri/xml/searchable.rb:198:in `evaluate'
	from /Users/cody/.rvm/gems/ruby-2.2.6@quickbooks-ruby/gems/nokogiri-1.8.2/lib/nokogiri/xml/searchable.rb:198:in `xpath_impl'
	from /Users/cody/.rvm/gems/ruby-2.2.6@quickbooks-ruby/gems/nokogiri-1.8.2/lib/nokogiri/xml/searchable.rb:179:in `xpath_internal'
	from /Users/cody/.rvm/gems/ruby-2.2.6@quickbooks-ruby/gems/nokogiri-1.8.2/lib/nokogiri/xml/searchable.rb:154:in `xpath'
	from (irb):7
	from /Users/cody/.rvm/rubies/ruby-2.2.6/bin/irb:11:in `<main>'
2.2.6 :008 >
```

So based on this I can only see your exception happening IF we are calling `Nokogiri#remove_namespaces!` -- which we are not. So I'm baffled

**Comment by djskim27** [comment]
I'm baffled too.. I'm guessing that I'm the only one experiencing this error so maybe something is off in my configurations for this app.

I decided to refactor and use QBO instead. I really wanted to stick with quickbooks-ruby but my client wanted his quickbooks stuff running asap. Thanks for all your help though @ruckus

**Comment by tpiekos** [comment]
I ran into this same issue as well, I was able to isolate it to Faraday not actually using the registered middleware for gzip decompression that's been added despite being on the correct branch, commit, etc. I have no idea if it's unique to my configuration/app/setup, but what seems to have fixed it for me was passing in a block when the OAuth2 client is initialized to configure Faraday like this -

```ruby
::QB_OAUTH2_CONSUMER = OAuth2::Client.new(QB_OAUTH_CONSUMER_KEY, QB_OAUTH_CONSUMER_SECRET, oauth_params) do |builder|
  builder.use :gzip
  builder.request :url_encoded
  builder.adapter Faraday.default_adapter
end
```

**Comment by ruckus** [comment]
@tpiekos your last comment was helpful. It also jogged my memory on something I had seen when @djskim27 posted a screenshot from earlier.

In his screenshot: https://i.imgur.com/dKfjYD1.png

At the very top there is a line about `Cannot rebuild a locked connection...`

Which we know is in the `verify_multipart!` method which rebuilds the Faraday connection with both *multipart* AND *gzip* support.

The problem is when that message is being logged the method is in the process of early-exiting and thus never rebuilds the connection *with gzip support*.

I've pushed an update which basically always rebuilds the connection when a new Access Token is assigned.

Thanks for the feedback and comments!

**Comment by fabrouy** [comment]
I was getting the exact same issue but only on the sandbox environment. In production everything was running fine.

It is not the first time Intuit made changes without letting everyone know. Remember this? https://github.com/ruckus/quickbooks-ruby/issues/409

And my guess is that when they make changes not everyone of us gets affected because they do not deploy to all their user base at once. So it's harder to debug. Anyway just wanted to let you know I was experiencing the same problem.

**Comment by gregawoods** [comment]
I'm getting the same issue in sandbox and production. Spent some time debugging and came to the same conclusion regarding gzip.

One problem I have now is I'm trying to update my gemfile to pull down the fix, but I can't because roxml has been updated to 4.0. https://github.com/ruckus/quickbooks-ruby/commit/85c77b5103e3b942c0506239491c6c7b1db22caf#diff-da82320b49f9ab083076f15d5adb8505R19

My roxml is locked in to 3.x because I use the [authorizenet](https://rubygems.org/gems/authorizenet/versions/1.9.4) gem.

Any way I can fix this in my code without updating roxml?

**Comment by gregawoods** [comment]
Update (should have tried this first): Pulling down the "future" branch of authorizenet resolved the issue for me, as roxml 4.0 is supported there.

```ruby
gem 'authorizenet', git: 'https://github.com/AuthorizeNet/sdk-ruby.git', branch: 'future'
gem 'quickbooks-ruby', git: 'https://github.com/ruckus/quickbooks-ruby.git', branch: '389-oauth2'
```

I can confirm that the XML parsing error is also resolved for me when using the latest code. Thanks!
